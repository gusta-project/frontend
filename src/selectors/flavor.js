import { createSelector } from 'reselect';

export const getFlavor = state => state.flavor;

export const isLoaded = createSelector(getFlavor, flavor =>
  Boolean(flavor.loaded)
);

export const getError = createSelector(getFlavor, flavor => flavor.error);

export const getStash = createSelector(getFlavor, flavor => flavor.stash);
