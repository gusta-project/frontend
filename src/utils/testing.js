import React, { createElement } from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

/**
 * Creates a mock component which takes a set of props specified at runtime.
 *
 * @param {string} name The component/element name (in PascalCase)
 * @param {object} props An object containing props to setup
 * @return {object} Mock component with specified name and props
 */
export const mockComponent = (name, props = {}) => () =>
  createElement(name, props, props.children);

/**
 * Creates a mock component which will expose its props for snapshot testing purposes
 * @param {*} name The component/element name (in PascalCase)
 */
export const mockComponentWithProps = (name) => (props) =>
  createElement(name, props, props.children);

/**
 * Wraps a React component in a <MemoryRouter> suitable for testing.
 *
 * @param {Component} WrappedComponent React component to wrap
 */
export const withMemoryRouter = (WrappedComponent, routerProps = {}) => (
  props
) => (
  <MemoryRouter {...routerProps}>
    <WrappedComponent {...props} />
  </MemoryRouter>
);

/**
 * Wraps a React component in a <Provider> suitable for testing.
 *
 * @param {Component} WrappedComponent React component to wrap
 * @param {object} store Mock store
 */
export const withProvider = (WrappedComponent, store) => (props) => (
  <Provider store={store}>
    <WrappedComponent {...props} />
  </Provider>
);
