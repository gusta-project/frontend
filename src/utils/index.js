import dayjs from 'dayjs';

import { getApplication, getAuthorization } from 'selectors/application';

/* eslint-disable react/display-name, react/prop-types */

/**
 * Creates an object containing action constants namespaced under the specified reducer.
 *
 * @param {string} reducer The name of the reducer or module
 * @param {string[]} actions A list of action names
 * @return {object} Object with action name keys and full action string values
 */
export function buildActions(reducer, actions) {
  const result = {};

  for (const action of actions) {
    result[action] = `${reducer}/${action}`;
  }

  return result;
}

/**
 * Interpolate strings from the `params` object into the `url`.
 *
 * @param {object} endpoint Object with `url` and optional `params` object
 */
export const buildUrl = (endpoint) => {
  // this is defined by webpack.DefinePlugin
  const baseUrl = API_URL;
  const { url, params } = endpoint;

  if (!params) {
    return `${baseUrl}${url}`;
  }

  return Object.entries(params).reduce((resultUrl, entry) => {
    const [key, value] = entry;

    return resultUrl.replace(`{${key}}`, value);
  }, `${baseUrl}${url}`);
};

/**
 * Gets the initial Redux state from localStorage and the source code defaults.
 */
export const getInitialState = () => {
  // this needs to be require()d because an import results in a
  // circular dependency
  const { initialState } = require('reducers');

  try {
    const [accessToken, expiration] = [
      localStorage.getItem('accessToken'),
      localStorage.getItem('expiration')
    ];

    if (!accessToken) {
      return initialState;
    }

    const expirationDate = dayjs(JSON.parse(expiration));

    if (!expirationDate.isValid() || expirationDate.isBefore(dayjs())) {
      localStorage.removeItem('accessToken');
      localStorage.removeItem('expiration');
      return initialState;
    }

    return {
      ...initialState,
      application: {
        ...getApplication(initialState),
        authorization: {
          ...getAuthorization(initialState),
          accessToken: JSON.parse(accessToken),
          expiration: expirationDate
        }
      }
    };
  } catch {
    return initialState;
  }
};
