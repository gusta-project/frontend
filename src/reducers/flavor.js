import { buildActions } from 'utils';

export const types = buildActions('flavor', [
  'REQUEST_STASH',
  'REQUEST_STASH_SUCCESS',
  'REQUEST_STASH_FAILURE'
]);

const requestStash = () => ({
  type: types.REQUEST_STASH
});

const requestStashSuccess = stash => ({
  type: types.REQUEST_STASH_SUCCESS,
  stash
});

const requestStashFailure = error => ({
  type: types.REQUEST_STASH_FAILURE,
  error
});

export const actions = {
  requestStash,
  requestStashSuccess,
  requestStashFailure
};

export const initialState = {
  loaded: false,
  error: null,
  stash: []
};

export const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case types.REQUEST_STASH_SUCCESS:
      return {
        ...state,
        stash: action.stash,
        loaded: true
      };
    case types.REQUEST_STASH_FAILURE:
      return {
        ...state,
        error: action.error
      };
    default:
      return state;
  }
};
