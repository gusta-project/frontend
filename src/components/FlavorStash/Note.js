import PropTypes from 'prop-types';
import React, { Fragment, useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import unified from 'unified';
import parse from 'remark-parse';
import remark2react from 'remark-react';
import { Form as FinalForm, Field } from 'react-final-form';
import { Button, ButtonGroup, Col, Form, InputGroup } from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { actions as noteActions } from 'reducers/note';
import { getFlavorNote, isLoading } from 'selectors/note';

export default function Note({ flavorId, userId }) {
  Note.propTypes = {
    flavorId: PropTypes.string.isRequired,
    userId: PropTypes.string.isRequired
  };

  const dispatch = useDispatch();
  const loading = useSelector(isLoading);
  const note = useSelector(getFlavorNote);
  const [editing, setEditing] = useState(false);

  useEffect(() => {
    dispatch(noteActions.requestNote({ flavorId }));
  }, [dispatch]);

  const handleToggleEditing = () => {
    setEditing(!editing);
  };

  const addNoteIcon = (id) => {
    return (
      <Fragment>
        {!editing && (
          <Button
            className="button-animation"
            size="sm"
            onClick={() => handleToggleEditing(id)}
          >
            <FontAwesomeIcon icon="plus" size="sm" title="Add Flavor Note" />
          </Button>
        )}
      </Fragment>
    );
  };

  const editNoteIcon = (id) => {
    return (
      <Fragment>
        {!editing && (
          <Button
            className="button-animation"
            size="sm"
            onClick={() => handleToggleEditing(id)}
          >
            <FontAwesomeIcon icon="pen" size="sm" title="Edit Flavor Note" />
          </Button>
        )}
      </Fragment>
    );
  };

  const handleNoteSubmit = (values) => {
    const { update } = values;
    const noteEdit = values.note;

    if (!noteEdit) {
      dispatch(noteActions.deleteNote(values));
    } else if (update === true) {
      dispatch(noteActions.updateNote(values));
    } else {
      dispatch(noteActions.createNote(values));
    }

    handleToggleEditing();
  };

  const noteEditor = () => {
    const update = Boolean(note?.[flavorId]?.note);

    return (
      <FinalForm
        onSubmit={handleNoteSubmit}
        initialValues={{
          flavorId,
          note: note?.[flavorId]?.note,
          update,
          userId
        }}
        render={({ handleSubmit, submitting, values }) => (
          <Form noValidate onSubmit={handleSubmit}>
            <Form.Row>
              <Field name="note">
                {({ input }) => (
                  <Form.Group as={Col}>
                    <InputGroup>
                      <Form.Control
                        {...input}
                        as="textarea"
                        placeholder="Flavor Note"
                        rows="10"
                      />
                    </InputGroup>
                  </Form.Group>
                )}
              </Field>
            </Form.Row>
            {values?.note ? (
              <Form.Row>
                <h4>Preview:</h4>
                <Col md="12">
                  {
                    /* eslint-disable-next-line no-sync */
                    unified()
                      .use(parse)
                      .use(remark2react)
                      .processSync(values.note).result
                  }
                </Col>
              </Form.Row>
            ) : null}
            <Form.Row>
              <Form.Group>
                <ButtonGroup>
                  <Button
                    className="button-animation"
                    variant="primary"
                    type="submit"
                    disabled={submitting}
                  >
                    <FontAwesomeIcon icon="save" size="sm" />
                    &nbsp;<span>Save</span>
                  </Button>
                  <Button
                    onClick={(e) => handleToggleEditing(false, e)}
                    className="button-animation button--cancel"
                  >
                    <span>Cancel</span>
                  </Button>
                </ButtonGroup>
              </Form.Group>
            </Form.Row>
          </Form>
        )}
      />
    );
  };

  if (loading !== flavorId) {
    return (
      <Fragment>
        <h3>
          {note[flavorId] ? editNoteIcon(flavorId) : addNoteIcon(flavorId)}{' '}
          Flavor Notes
        </h3>
        {!editing && note[flavorId] ? (
          <Fragment>
            {
              /* eslint-disable-next-line no-sync */
              unified()
                .use(parse)
                .use(remark2react)
                .processSync(note[flavorId].note).result
            }
          </Fragment>
        ) : null}
        {editing ? noteEditor() : null}
      </Fragment>
    );
  } else {
    return <Fragment>Loading...</Fragment>;
  }
}
