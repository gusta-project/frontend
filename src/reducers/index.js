import { combineReducers } from 'redux';

import {
  reducer as application,
  initialState as applicationState
} from './application';
import {
  reducer as dashboard,
  initialState as dashboardState
} from './dashboard';
import { reducer as flavor, initialState as flavorState } from './flavor';
import { reducer as flavors, initialState as flavorsState } from './flavors';
import { reducer as note, initialState as noteState } from './note';
import { reducer as profile, initialState as profileState } from './profile';
import { reducer as recipe, initialState as recipeState } from './recipe';
import { reducer as roles, initialState as rolesState } from './roles';
import { reducer as toast, initialState as toastState } from './toast';
import { reducer as users, initialState as usersState } from './users';
import { getInitialState } from 'utils';

export const initialState = {
  application: applicationState,
  dashboard: dashboardState,
  flavor: flavorState,
  flavors: flavorsState,
  note: noteState,
  profile: profileState,
  recipe: recipeState,
  roles: rolesState,
  toast: toastState,
  users: usersState
};

const reducer = combineReducers({
  application,
  dashboard,
  flavor,
  flavors,
  note,
  profile,
  recipe,
  roles,
  toast,
  users
});

export default function (state, action) {
  if (action.type === 'application/INIT_APP') {
    state = getInitialState();
  }

  return reducer(state, action);
}
