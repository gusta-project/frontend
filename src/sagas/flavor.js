import { all, call, put, takeLatest, select, take } from 'redux-saga/effects';

import request from 'utils/request';
import { getUser } from 'selectors/application';
import { actions, types } from 'reducers/flavor';
import { actions as appActions, types as appTypes } from 'reducers/application';

function* requestStashWorker() {
  try {
    let user = yield select(getUser);

    if (user === null) {
      yield put(appActions.requestCurrentUser());
      yield take([
        appTypes.REQUEST_CURRENT_USER_SUCCESS,
        appTypes.REQUEST_CURRENT_USER_FAILURE
      ]);
      user = yield select(getUser);
    }

    const endpoint = {
      url: `/user/${user.id}/flavors`,
      method: 'GET'
    };
    const result = yield call(request.execute, { endpoint });

    if (result.success) {
      const { data } = result.response;

      if (!data) {
        return yield put(actions.requestStashSuccess([]));
      }

      yield put(actions.requestStashSuccess(data));
    } else if (result.error) {
      throw result.error;
    } else {
      throw new Error('Request failed for an unspecified reason!');
    }
  } catch (error) {
    // eslint-disable-next-line
    console.dir(error);
    yield put(actions.requestStashFailure(error));
  }
}

export const workers = {
  requestStashWorker
};

function* requestStashWatcher() {
  yield takeLatest(types.REQUEST_STASH, requestStashWorker);
}

export const watchers = {
  requestStashWatcher
};

export default function* saga() {
  yield all(Object.values(watchers).map(watcher => watcher()));
}
