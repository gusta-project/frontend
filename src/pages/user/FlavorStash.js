import dayjs from 'dayjs';
import { Helmet } from 'react-helmet';
import React, { Fragment, useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { Form as FinalForm, Field } from 'react-final-form';
import {
  Button,
  ButtonGroup,
  Card,
  Col,
  Container,
  Form,
  InputGroup,
  Row
} from 'react-bootstrap';
import ToggleButton from 'components/ToggleButton/ToggleButton';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { actions as flavorActions } from 'reducers/flavor';
import { getStash, isLoaded } from 'selectors/flavor';
import Note from 'components/FlavorStash/Note';

export default function FlavorStash() {
  const dispatch = useDispatch();
  const stash = useSelector(getStash);
  const stashLoaded = useSelector(isLoaded);

  const [expanded, setExpanded] = useState({});
  const [editingStash, setEditing] = useState(false);
  const [removed, setRemoved] = useState({});
  const [usage, setUsage] = useState({});

  useEffect(() => {
    if (!stashLoaded) {
      dispatch(flavorActions.requestStash());
    }
  }, [stashLoaded, dispatch]);

  const date = (d) => {
    return dayjs(d).format('YYYY-MM-DD');
  };

  const addToStash = (id) => {
    dispatch(flavorActions.addStash({ id }));
    removed[id] = false;
    setRemoved({ ...removed });
  };

  const removeFromStash = (id) => {
    dispatch(flavorActions.removeStash({ id }));
    removed[id] = true;
    setRemoved({ ...removed });
  };

  const expandIcon = (isExpanded) => {
    return (
      <FontAwesomeIcon
        rotation={expanded ? 90 : 270}
        icon="chevron-left"
        className={`mr-2 ${isExpanded ? 'arrow--up' : 'arrow--down'}`}
      />
    );
  };

  const stashButton = (id, has) => {
    return (
      <ToggleButton
        value={has}
        onClick={has ? () => removeFromStash(id) : () => addToStash(id)}
        title={has ? 'Remove from Stash' : 'Add to Stash'}
        variant="check"
      />
    );
  };

  const handleStashEditor = (id) => {
    setEditing(id ? id : false);
  };

  const editIcon = (id) => {
    return (
      <Fragment>
        {!editingStash ? (
          <Button
            className="button-animation"
            size="sm"
            onClick={(e) => handleStashEditor(id, e)}
          >
            <FontAwesomeIcon icon="pen" size="sm" title="Edit Details" />
          </Button>
        ) : null}
      </Fragment>
    );
  };

  const handleStashSubmit = (values) => {
    const { flavorId, minMillipercent, maxMillipercent } = values;

    dispatch(flavorActions.updateStash(values));

    usage[flavorId] = {
      minMillipercent,
      maxMillipercent
    };
    setEditing(false);
    setUsage({ ...usage });
  };

  const expandFlavor = (flavor) => {
    const { flavorId } = flavor;

    expanded[flavorId] = !expanded?.[flavorId];

    setExpanded({ ...expanded });
  };

  const stashEditor = (flavor) => {
    const { flavorId, maxMillipercent, minMillipercent } = flavor;

    return (
      <FinalForm
        onSubmit={handleStashSubmit}
        initialValues={{
          flavorId,
          maxMillipercent:
            usage?.[flavorId]?.maxMillipercent || maxMillipercent / 1000,
          minMillipercent:
            usage?.[flavorId]?.minMillipercent || minMillipercent / 1000
        }}
        render={({ handleSubmit, submitting }) => (
          <Form noValidate onSubmit={handleSubmit}>
            <h3>Usage Range:</h3>
            <Form.Row>
              <Field name="minMillipercent">
                {({ input }) => (
                  <Form.Group as={Col}>
                    <InputGroup>
                      <InputGroup.Prepend>
                        <InputGroup.Text id="inputGroupPrepend">
                          Min
                        </InputGroup.Text>
                      </InputGroup.Prepend>
                      <Form.Control {...input} type="number" step="0.1" />
                      <InputGroup.Text>%</InputGroup.Text>
                    </InputGroup>
                  </Form.Group>
                )}
              </Field>
              <Field name="maxMillipercent">
                {({ input }) => (
                  <Form.Group as={Col}>
                    <InputGroup>
                      <InputGroup.Prepend>
                        <InputGroup.Text id="inputGroupPrepend">
                          Max
                        </InputGroup.Text>
                      </InputGroup.Prepend>
                      <Form.Control {...input} type="number" step="0.1" />
                      <InputGroup.Text>%</InputGroup.Text>
                    </InputGroup>
                  </Form.Group>
                )}
              </Field>
              <Field name="flavorId">
                {({ input }) => <Form.Control {...input} type="hidden" />}
              </Field>
            </Form.Row>
            <Form.Row>
              <Form.Group>
                <ButtonGroup>
                  <Button
                    className="button-animation"
                    variant="primary"
                    type="submit"
                    disabled={submitting}
                  >
                    <FontAwesomeIcon icon="save" size="sm" />
                    &nbsp;<span>Save</span>
                  </Button>
                  <Button
                    onClick={(e) => handleStashEditor(false, e)}
                    className="button-animation button--cancel"
                    variant="danger"
                  >
                    <span>Cancel</span>
                  </Button>
                </ButtonGroup>
              </Form.Group>
            </Form.Row>
          </Form>
        )}
      />
    );
  };

  return (
    <Container>
      <Helmet title="Your Flavor Stash" />
      <Row className="text-center">
        <Col>
          <h1>Flavor Stash</h1>
        </Col>
      </Row>
      <Row>
        <Col>
          {stash.map((flavor, index) => {
            const minMillipercent = flavor.minMillipercent / 1000;
            const maxMillipercent = flavor.maxMillipercent / 1000;
            const milliPercents = usage[flavor.flavorId]
              ? `${usage[flavor.flavorId].minMillipercent} - ${
                  usage[flavor.flavorId].maxMillipercent
                }`
              : `${minMillipercent} - ${maxMillipercent}`;

            return (
              <Card key={index} className="mb-1">
                <Card.Body
                  className="border-bottom cursor--pointer"
                  onClick={(e) => expandFlavor(flavor, e)}
                >
                  <Row>
                    <Col>
                      <h3
                        id={`flavor-${flavor.flavorId}`}
                        className={
                          expanded[flavor.flavorId] ? 'text--primary' : null
                        }
                      >
                        {expandIcon(Boolean(expanded[flavor.flavorId]))}{' '}
                        {flavor.Flavor.name} ({flavor.Flavor.Vendor.code})
                      </h3>
                    </Col>
                    {!expanded[flavor.flavorId] && (
                      <Col className="text-center">
                        <h4>
                          <strong>Use:</strong> {milliPercents}%
                        </h4>
                      </Col>
                    )}
                    <Col className="text-right">
                      <h5>Added: {date(flavor.created)}</h5>
                    </Col>
                  </Row>
                </Card.Body>

                {expanded[flavor.flavorId] && (
                  <Fragment>
                    <Card.Body>
                      <Card.Title>{flavor.Flavor.Vendor.name}</Card.Title>

                      <Row>
                        <Col>
                          <strong>Density:</strong>{' '}
                          {flavor.Flavor.density || 'Not Available'}
                        </Col>
                        <Col>
                          {editIcon(flavor.flavorId)} <strong>Use:</strong>{' '}
                          {milliPercents}%
                        </Col>
                      </Row>

                      {editingStash === flavor.flavorId
                        ? stashEditor(flavor)
                        : ''}
                      <hr />
                      <Note flavorId={flavor.flavorId} userId={flavor.userId} />
                    </Card.Body>
                    <Card.Footer>
                      <span className="float-left">ID: {flavor.flavorId}</span>
                      <span className="float-right">
                        {stashButton(
                          flavor.flavorId,
                          Boolean(!removed[flavor.flavorId])
                        )}
                      </span>
                    </Card.Footer>
                  </Fragment>
                )}
              </Card>
            );
          })}
          {!stash[0] && (
            <Col className="text-center">
              Navigate to{' '}
              <Link to="/flavors" title="Flavors">
                Flavors
              </Link>{' '}
              to add to your Flavor Stash.
              <p>No flavors in stash.</p>
            </Col>
          )}
        </Col>
      </Row>
    </Container>
  );
}
