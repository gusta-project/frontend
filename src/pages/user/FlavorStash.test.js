import { render, waitFor } from '@testing-library/react';
import React from 'react';
import { useDispatch } from 'react-redux';
import configureStore from 'redux-mock-store';

import FlavorStash from './FlavorStash';
import { actions as flavorActions } from 'reducers/flavor';
import { withMemoryRouter, withProvider } from 'utils/testing';
import { initialState } from 'reducers';

jest.mock('react-redux', () => {
  const dispatch = jest.fn();

  return {
    ...jest.requireActual('react-redux'),
    useDispatch: jest.fn(() => dispatch)
  };
});

describe('<FlavorStash />', () => {
  const mockStore = configureStore();
  const emptyStore = mockStore(initialState);
  const stashStore = mockStore({
    ...initialState,
    flavor: {
      ...initialState.flavor,
      stash: [
        {
          created: '2020-01-01T00:00:00',
          Flavor: {
            id: 1,
            name: 'Sweet Cream',
            Vendor: { code: 'CAP' }
          }
        },
        {
          created: '2020-01-01T00:00:00',
          Flavor: {
            id: 2,
            name: 'Tatanka',
            Vendor: { code: 'FLV' }
          }
        }
      ]
    }
  });

  it('can render', () => {
    const ConnectedStash = withProvider(FlavorStash, stashStore);
    const { asFragment } = render(<ConnectedStash />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('can render with an empty stash', async () => {
    const dispatch = useDispatch();
    const ConnectedStash = withMemoryRouter(
      withProvider(FlavorStash, emptyStore)
    );
    const { asFragment } = render(<ConnectedStash />);

    await waitFor(() => {
      expect(dispatch).toHaveBeenCalledWith(flavorActions.requestStash());
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
