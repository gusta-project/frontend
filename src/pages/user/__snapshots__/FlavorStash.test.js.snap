// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlavorStash /> can render 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="text-center row"
    >
      <div
        class="col"
      >
        <h1>
          Flavor Stash
        </h1>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="mb-1 card"
        >
          <div
            class="border-bottom cursor--pointer card-body"
          >
            <div
              class="row"
            >
              <div
                class="col"
              >
                <h3
                  id="flavor-undefined"
                >
                  <i
                    class="fa chevron-left"
                  />
                   Sweet Cream (CAP)
                </h3>
              </div>
              <div
                class="text-center col"
              >
                <h4>
                  <strong>
                    Use:
                  </strong>
                   NaN - NaN%
                </h4>
              </div>
              <div
                class="text-right col"
              >
                <h5>
                  Added: 2020-01-01
                </h5>
              </div>
            </div>
          </div>
        </div>
        <div
          class="mb-1 card"
        >
          <div
            class="border-bottom cursor--pointer card-body"
          >
            <div
              class="row"
            >
              <div
                class="col"
              >
                <h3
                  id="flavor-undefined"
                >
                  <i
                    class="fa chevron-left"
                  />
                   Tatanka (FLV)
                </h3>
              </div>
              <div
                class="text-center col"
              >
                <h4>
                  <strong>
                    Use:
                  </strong>
                   NaN - NaN%
                </h4>
              </div>
              <div
                class="text-right col"
              >
                <h5>
                  Added: 2020-01-01
                </h5>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<FlavorStash /> can render with an empty stash 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="text-center row"
    >
      <div
        class="col"
      >
        <h1>
          Flavor Stash
        </h1>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="text-center col"
        >
          Navigate to 
          <a
            href="/flavors"
            title="Flavors"
          >
            Flavors
          </a>
           to add to your Flavor Stash.
          <p>
            No flavors in stash.
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
